### PR TITLE
WebSockets for netty-cats

### PR DESCRIPTION
**Pull request by kciesielski**
This PR adds support for Web Socket endpoints in `tapir-netty-cats`, with all the sub-features like `autoPongOnPing`, `ignorePong`, `decodeCloseRequests`, `decodeCloseResponses`, and `autoPing`.

Some remarks:
* The logic for handling control frames (Close, pings/pongs) happens mostly in common tapir-netty code and should work for other Netty-based backends.
* The Cats Effect - specific part is `Fs2StreamCompatible`, and `WebSocketPipeProcessor`. Especially the latter, which creates a Reactive Streams `Processor[Req, Resp]` from user-provided `fs2.Pipe[F, Req, Resp]`.
* The `WebSocketPipeProcessor` has a workaround to properly handle exceptions thrown in the pipeline. It needs to prevent underlying cancelation mechanisms and instead signal back an error to the Netty channel handler, so that the server properly sends a Close frame. Otherwise the channel closes silently, which also happens in some other servers, like Vert.X, so tests for error handling run only when `failingPipe` is set to true in `ServerWebSocketTests`.
* I also added tests for ping/pong handling, which are selectively disabled for some interpreters which don't support them.

In performance tests, latency is on par with our fastest backends (like http4s):

image

### CODE REVIEW

File: server/netty-server/cats/src/main/scala/sttp/tapir/server/netty/cats/NettyCatsServer.scala
Comment (adamw) [inline server/netty-server/cats/src/main/scala/sttp/tapir/server/netty/cats/NettyCatsServer.scala:33]: I guess we should also add the capability to `addEndpoint` above?

File: server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala
Comment (adamw) [inline server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala:73]: maybe move this to a constant?

File: server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala
Comment (adamw) [inline server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala:129]: is this really an error? that is, won't this occur when a client attempts to upgrade a "normal" endpoint to a websocket? (that would be client's error, but nothing to fix on the server-side)

File: server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala
Comment (adamw) [inline server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala:91]: maybe move this to a dedicated function? this seems to be "the" place where we handle the WS logic, while all the other branches handle error conditions

File: server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala
Comment (adamw) [inline server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala:121]: maybe extract these into methods as well, with names specifying what kind of error / misues occured, e.g. `serverErrorUnexpectedResponseContent` or `clientErrorNotAWSEndpoint` or sth like this. The main flow would read much nicer then

File: server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala
Comment (adamw) [reply server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala:129]: yes, it's definitely a client error, not a server one, so 4xx is the appropriate status code. Plus, we shouldn't `logger.error`, as that's not a server error - but misuse by the client

Comment (kciesielski) [reply server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala:129]: Fun fact: I checked on a few other servers, and they return a HTTP 200 response with normal content from the non-ws endpoint, totally ignoring that the request was a WS handshake.

Comment (kciesielski) [reply server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala]: Yes, checked that with vanilla akka-http and http4s.

Comment (kciesielski) [reply server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala]: But I'm convinced we should return HTTP 400 in such case.

Comment (adamw) [reply server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala]: Yes, this sounds reasonable :)

File: server/netty-server/cats/src/main/scala/sttp/tapir/server/netty/cats/internal/Fs2StreamCompatible.scala
Comment (adamw) [inline server/netty-server/cats/src/main/scala/sttp/tapir/server/netty/cats/internal/Fs2StreamCompatible.scala:74]: extract this to a helper method `sendCloseFramePromise`?

Comment (adamw) [reply server/netty-server/cats/src/main/scala/sttp/tapir/server/netty/cats/internal/Fs2StreamCompatible.scala:74]: or rename the value, `processorPromise` doesn't reveal too much about its purpose :)

Comment (adamw) [reply server/netty-server/cats/src/main/scala/sttp/tapir/server/netty/cats/internal/Fs2StreamCompatible.scala:74]: reading the code furhter, I guess this should just be `wsCompletedPromise`?

File: server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala
Comment (adamw) [inline server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala:25]: is this based on some cats-effects class? if so we should add a backlink

File: server/netty-server/src/main/scala/sttp/tapir/server/netty/NettyConfig.scala
Comment (adamw) [inline server/netty-server/src/main/scala/sttp/tapir/server/netty/NettyConfig.scala:64]: the comment for initPipeline needs an update, or formalising as a trait since the ChannelHandler parameters can be confusing. Or maybe pass a list of handlers?

File: server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/WebSocketControlFrameHandler.scala
Comment (adamw) [inline server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/WebSocketControlFrameHandler.scala:34]: will the upstream know then that the WS is closed?

File: server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/NettyServerHandler.scala
Comment (adamw) [inline server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/NettyServerHandler.scala:216]: shouldn't this be a 400 as well? would correspond to the case where a ws endpoint is called without the ws upgrade headers, no?

Comment (adamw) [reply server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/NettyServerHandler.scala:216]: actually maybe we should have tests, that both calling a WS endpoint normally, and calling with WS a normal endpoint end up with a 400

File: server/tests/src/main/scala/sttp/tapir/server/tests/ServerWebSocketTests.scala
Comment (adamw) [inline server/tests/src/main/scala/sttp/tapir/server/tests/ServerWebSocketTests.scala:207]: in fact we have a test for non-ws request to a ws endpoint above, so maybe this works, just would need extending with a status code check

still a test for a ws request to a non-ws endpoint could be useful

File: server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala
Comment (adamw) [inline server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala:76]: typo in handshake

File: server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala
Comment (adamw) [inline server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala:87]: so we don't need the constants after all? :)

File: server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala
Comment (adamw) [inline server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala:107]: won't this shadow a 404?

File: server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala
Comment (adamw) [inline server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala:111]: doesn't the lack of a body simply  mean that the endpoint was a "normal" one, but didn't specify a body output? So, is this a server error?

File: server/netty-server/src/main/scala/sttp/tapir/server/netty/NettyFutureServer.scala
Comment (adamw) [inline server/netty-server/src/main/scala/sttp/tapir/server/netty/NettyFutureServer.scala:76]: if we were to pass a list of handlers here, maybe we could skip this handler when websockets are not supported? (as in `Future`)

File: server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/WebSocketAutoPingHandler.scala
Comment (adamw) [inline server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/WebSocketAutoPingHandler.scala:18]: maybe move these to a ws package? could help to organize the code a bit :)

File: server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala
Comment (kciesielski) [reply server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala:87]: We need them for `pipeline.addAfter`, where you have to use names.

File: server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/WebSocketControlFrameHandler.scala
Comment (kciesielski) [reply server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/WebSocketControlFrameHandler.scala:34]: It may be indeed a good idea to close the channel explicitly here. Sending a CLOSE back probably makes only the client close its connection, which gives fine behavior in tests, but we shouldn't rely on that.

File: server/tests/src/main/scala/sttp/tapir/server/tests/ServerWebSocketTests.scala
Comment (kciesielski) [reply server/tests/src/main/scala/sttp/tapir/server/tests/ServerWebSocketTests.scala:207]: We have a test:
```scala
  testServer(
      endpoint
        .in(isWebSocket)
        .errorOut(stringBody)
        .out(stringWs),
      "non web-socket request"
    )(isWS => if (isWS) pureResult(stringEcho.asRight) else pureResult("Not a WS!".asLeft)) { (backend, baseUri) =>
```

does this mean that the user has to explicitly handle this?

File: server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/WebSocketControlFrameHandler.scala
Comment (adamw) [reply server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/WebSocketControlFrameHandler.scala:34]: I was thinking about the fs2 stream, that should shut down as well, will it? :)

File: server/tests/src/main/scala/sttp/tapir/server/tests/ServerWebSocketTests.scala
Comment (adamw) [reply server/tests/src/main/scala/sttp/tapir/server/tests/ServerWebSocketTests.scala:207]: Yes and no ;) This test probably needs improving / commenting: I think it might test for the case of a "dynamic" web socket output - that is, we can decide dynamically if we return a web socket, or not. 

The endpoint above works for both cases, deciding in the server logic what type of output to produce. Using `isWebSocket` is probably just me being lazy, and the test should use a query parameter or sth like this, and be split into four: with/without WS upgrade headers, with/without the "correct" query parameter value.

This would also test for the case when we want to do some validation of the request, before returning a web socket - e.g, testing security - which case we want to return a nice error to the user

File: server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala
Comment (kciesielski) [reply server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala:111]: Yes, I just checked and an empty body endpoint would trigger this. What's worse, a 404 would also end up here. I'll fix this.

File: server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala
Comment (kciesielski) [reply server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala:25]: No, I wrote it from scratch, is that what you mean?

Comment (adamw) [reply server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala:25]: ah ok :) the comment "By using the Netty ..." looked formatted differently than our other comments so I thought that's taken from somewhere - sorry :)

File: server/netty-server/cats/src/main/scala/sttp/tapir/server/netty/cats/internal/WebSocketPipeProcessor.scala
Comment (kciesielski) [reply server/netty-server/cats/src/main/scala/sttp/tapir/server/netty/cats/internal/WebSocketPipeProcessor.scala:32]: It's set in `onSubscribe` which can be run from another thread. Netty internal reactive publisher and subscriber follow the same approach. They set other internal values without volatile, because they are updated/read only in onNext/onError etc, which are run always within Netty eventLoopExecutor, which guarantees the same thread (at least that's my understanding). Values set in `onSubscribe` don't have this guarantee.

Comment (adamw) [reply server/netty-server/cats/src/main/scala/sttp/tapir/server/netty/cats/internal/WebSocketPipeProcessor.scala:32]: ok, maybe add this as a comment for your future self :)

File: server/netty-server/cats/src/main/scala/sttp/tapir/server/netty/cats/internal/WebSocketPipeProcessor.scala
Comment (ghik) [inline server/netty-server/cats/src/main/scala/sttp/tapir/server/netty/cats/internal/WebSocketPipeProcessor.scala:40]: Can't we just initialize it in constructor and turn `@volatile var` into a `val`?

File: server/netty-server/cats/src/main/scala/sttp/tapir/server/netty/cats/internal/WebSocketPipeProcessor.scala
Comment (ghik) [inline server/netty-server/cats/src/main/scala/sttp/tapir/server/netty/cats/internal/WebSocketPipeProcessor.scala:102]: Cosmetic: use `cancel()` (with parens) as it's a side-effectful invocation.

Comment (ghik) [reply server/netty-server/cats/src/main/scala/sttp/tapir/server/netty/cats/internal/WebSocketPipeProcessor.scala:102]: Actually, the `subscription` is a `NonCancelingSubscription` so this invocation does nothing.

BTW, why is it non-canceling?

File: server/netty-server/src/main/scala/sttp/tapir/server/netty/NettyConfig.scala
Comment (adamw) [inline server/netty-server/src/main/scala/sttp/tapir/server/netty/NettyConfig.scala:134]: it doesn't matter that the handlers are now not named?

Comment (kciesielski) [reply server/netty-server/src/main/scala/sttp/tapir/server/netty/NettyConfig.scala:134]: Names are useful if you want to explicitly add other handlers before or after specific handlers. There's no such need in case of these two.

File: server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ws/ReactiveWebSocketHandler.scala
Comment (adamw) [inline server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ws/ReactiveWebSocketHandler.scala:123]: [minor] maybe we can simply .transform the original future (without the .map in :116), and if it's a none, return a 404

File: server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ws/ReactiveWebSocketHandler.scala
Comment (adamw) [inline server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ws/ReactiveWebSocketHandler.scala:138]: one problem here: if a regular endpoint is hit with WS headers, it might end up with a 500 and an error log - so I'm not sure if we shouldn't run regular processing in this case

Comment (kciesielski) [reply server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ws/ReactiveWebSocketHandler.scala:138]: We do run regular processing in line 133 above, if the interpreter returns a standard, non-streaming, non-file response.

File: server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/package.scala
Comment (ghik) [inline server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/package.scala:33]: Cosmetic: my convention is always to declare constants as `final val`s. This generates a slightly leaner bytecode and allows the compiler to do constant folding. You can also use these constants in Java annotations.

File: server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/package.scala
Comment (ghik) [inline server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/package.scala:16]: Minor: why not declare it as value class?

File: server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ws/ReactiveWebSocketHandler.scala
Comment (adamw) [reply server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ws/ReactiveWebSocketHandler.scala:138]: Yes, but there's still an easily achievable situation where a client error (calling any endpoint - including streaming/file ones) with the incorrect headers results in a server error

maybe this can even be considered a security problem? as you can easily spam a server with `logger.error` (which might trigger monitoring)

File: server/tests/src/main/scala/sttp/tapir/server/tests/ServerWebSocketTests.scala
Comment (adamw) [inline server/tests/src/main/scala/sttp/tapir/server/tests/ServerWebSocketTests.scala:30]: nice, we now have much better coverage :) 👍 (across all backends)

File: server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/package.scala
Comment (kciesielski) [reply server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/package.scala:16]: I agree it's a good idea to make it a value class, but what's the value added in making the `val` private?

Comment (ghik) [reply server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/package.scala:16]: Why would you want to make the wrapped value to become an extension member? Do you ever want to call `someChannelFuture.cf`?

File: server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ws/ReactiveWebSocketHandler.scala
Comment (kciesielski) [reply server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ws/ReactiveWebSocketHandler.scala:138]: Ah ok, I get it now. How about changing this 500 to a 400 without logging then?

Comment (adamw) [reply server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ws/ReactiveWebSocketHandler.scala:138]: Ah yes sounds good. As long as we don't leak any resources? (connection counters, open files, open streams?)

File: server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/package.scala
Comment (kciesielski) [reply server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/package.scala:16]: There's a cost of increased verbosity in such a defensive approach, which isn't always worthwhile. If the value added is just "preventing possibility of calling `someChannelFuture.cf`" it doesn't seem to be any risk and I would lean towards keeping lower verbosity.

Comment (ghik) [reply server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/package.scala:16]: Whatever, that's minor, but I always thought that

```scala
implicit class Extensions(private val thing: Thing) extends AnyVal {
  // extensions
}
```

is one of the most standard idioms in Scala. You can see it all over the standard library etc.
Adding a single keyword doesn't look like verbosity to me.

Comment (kciesielski) [reply server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/package.scala:16]: I agree that the standalone `val` has no point (it's copied code), so I'm happy to make the class an `AnyVal`. As for verbosity, I think it's beneficial to have a bit less code, without losing/risking anything. However, a standard practice of making val  private here is also OK for me and it looks like Tapir has a few such cases already, so I'll follow the convention here as well. I'll also apply this to `NettyServerHandler` where it was copied from.

Comment (ghik) [reply server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/package.scala:16]: > without losing/risking anything
If this was a part of public library API with backwards compatibility contract, it would be much worse - the accidentally exposed field could never be removed or renamed without a proper deprecation cycle. That's why I think it's better to be as conservative as possible with what is exposed as public.

Comment (kciesielski) [reply server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/package.scala:16]: Good point with the binary compatibility. These particular internals aren't meant become public API, but sometimes stuff gets copy-pasted (as we're seeing here :) and it may indeed be better to stay a bit more defensive just in case.

File: server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ws/ReactiveWebSocketHandler.scala
Comment (kciesielski) [reply server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ws/ReactiveWebSocketHandler.scala:138]: Actually an open file or a stream (Publisher[]) might get leaked here if not handled. I'll take that into account, thanks!

File: server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/NettyServerHandler.scala
Comment (kciesielski) [inline server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/NettyServerHandler.scala:324]: Class inner implicit classes cannot be AnyVals

**Comment by kciesielski** [comment]
Hmm such extractions require further refactoring, because response handlers are internal implicit classes referencing fields of `NettyServerHandler` (like request, isShuttingDown, etc.).

**Comment by adamw** [comment]
@kciesielski let's skip them then ;)

**Comment by adamw** [comment]
otherwise this PR will end up being gigantic feature + refactor

**Comment by kciesielski** [comment]
I'm totally on the same page on this :grin

**Comment by kciesielski** [comment]
Thank @ghik, your comments were helpful :) 
@adamw waiting for your approval then.

**Review by ghik — APPROVED** [review]

**Comment by adamw** [comment]
I think in the end code is even simpler now. Great work - let's ship it :)

**Pull request merged**
